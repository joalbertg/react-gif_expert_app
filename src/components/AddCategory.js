import React, {useState} from 'react';
import PropTypes from 'prop-types';

const AddCategory = ({setCategories}) => {
  const [value, setValue] = useState('');
  const handleAdd = event => {
    setValue(event.target.value);
  }

  const handleSubmit = event => {
    event.preventDefault();

    if(value.trim().length > 0) {
      setCategories(categories => [...categories, value]);
      setValue('');
    }
  }

  return(
    <form onSubmit={handleSubmit}>
      <input
        type="text"
        value={value}
        onChange={handleAdd}
        placeholder='Add here...'
      />
    </form>
  );
}

AddCategory.propTypes = {
  setCategories: PropTypes.func.isRequired
}

export default AddCategory;

