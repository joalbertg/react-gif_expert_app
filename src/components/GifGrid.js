import React from 'react';

import { useFetchGifs } from '../hooks/useFetchGifs';

import GifGridItem from './GifGridItem';

const GifGrid = ({category}) => {
  const {data: gifs, loading} = useFetchGifs(category);

  return(
    <div className="card-grid">
      <h3 className="animate__animated animate__fadeIn">{category}</h3>
      { loading && <p className="loading animate__animated animate__flash">Loading</p> }
      {
        gifs.map(gif => (
          <GifGridItem
            key={gif.id}
            {...gif}
          />)
        )
      }
    </div>
  );
}

export default GifGrid;

