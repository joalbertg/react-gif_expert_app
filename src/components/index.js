export { default as AddCategory } from './AddCategory';
export { default as ListCategories } from './ListCategories';
export { default as GifGrid } from './GifGrid';

