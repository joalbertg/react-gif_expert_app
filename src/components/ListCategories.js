import React from 'react';

import GifGrid from './GifGrid';

const ListCategories = ({categories}) => {
  return(
    <>
      { categories ?
          categories.map(category => (
            <GifGrid
              key={category}
              category={category}
            />
          ))
        :
        <h3>No categories to display.</h3>
      }
    </>
  );
}

export default ListCategories;

