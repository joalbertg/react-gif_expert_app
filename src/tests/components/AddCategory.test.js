import React from 'react';
import { shallow } from 'enzyme';

import { AddCategory } from '../../components';

describe('Tests AddCategory component', () => {
  const setCategories = jest.fn();
  let wrapper = shallow(<AddCategory setCategories={setCategories} />);

  beforeEach(() => {
    jest.clearAllMocks();
    wrapper = shallow(<AddCategory setCategories={setCategories} />);
  });

  test('Should display correctly', () => {
    const section = wrapper.find('section');
    const fontAwesomeIcon = wrapper.find('FontAwesomeIcon');
    const { iconName } = fontAwesomeIcon.prop('icon');

    expect(wrapper).toMatchSnapshot();
    expect(section.prop('id')).toBe('search');
    expect(section.prop('children').length).toBe(2);
    expect(iconName).toBe('search');
  });

  test('Should change the text box', () => {
    const input = wrapper.find('input');
    const value = 'Dragon Ball';

    input.simulate('change', { target: { value } });
    const currentValue = wrapper.find('input').prop('value');

    expect(currentValue).toBe(value);
  });

  test('Should not call setCategories or setValue with the submit event', () => {
    wrapper.find('form').simulate('submit', { preventDefault(){} });
    expect(setCategories).not.toHaveBeenCalled();
  });

  test('Should set value in "", setCategories must be called once and as a function', () => {
    const input = wrapper.find('input');
    const value = 'Dragon Ball';

    input.simulate('change', { target: { value } });
    wrapper.find('form').simulate('submit', { preventDefault(){} });
    const currentValue = wrapper.find('input').prop('value');

    expect(currentValue).toBe('');
    expect(setCategories).toHaveBeenCalled();
    expect(setCategories).toHaveBeenCalledTimes(1);
    //que sea evaluado con alguna funcion
    expect(setCategories).toHaveBeenCalledWith(expect.any(Function));
  });
});

