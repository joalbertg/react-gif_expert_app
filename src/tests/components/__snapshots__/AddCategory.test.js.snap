// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests AddCategory component Should display correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    onChange={[Function]}
    placeholder="Add here..."
    type="text"
    value=""
  />
</form>
`;
